src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp

```diff
@@ -816,6 +816,10 @@ std::pair<std::vector<String>, bool> ReplicatedMergeTreeSinkImpl<async_insert>::
                     NameSet unused;
                     /// if we found part in deduplication hashes part must exists on some replica
                     storage.checkPartChecksumsAndAddCommitOps(zookeeper, part, ops, existing_part_name, unused);
+
+                    /// We have to check that the block_id still exists to avoid a race condition with DROP_RANGE
+                    block_unlock_op_idx = ops.size();
```

Comment:
I understand that we should update the exception message in:
```
        else if (multi_code == Coordination::Error::ZNONODE && zkutil::getFailedOpIndex(multi_code, responses) == block_unlock_op_idx)
        {
            throw Exception(ErrorCodes::QUERY_WAS_CANCELLED,
                            "Insert query (for block {}) was cancelled by concurrent ALTER PARTITION", block_number_lock->getPath());
        }
```

And instead of `ALTER PARTITION` it should say "operation" or something more generic, right?

Reply:
> And instead of ALTER PARTITION it should say "operation" or something more generic, right?

It can be canceled by ALTER PARTITION or TRUNCATE (which is basically DROP PARTITION for all partitions). But a part can be removed from zk as a result of a merge, for example. And looks like in rare cases it may trigger this message too after my changes